src/impl.jl

```diff
@@ -0,0 +1,337 @@
+            end
+            hashfn = kw.second
+            check_valid_alt_hash_name(__source__, hashfn)
+        elseif kw.first === :fields
+            if !(kw.second isa Tuple) || !all(f isa Symbol for f in kw.second)
+                error("$(__source__.file):$(__source__.line): invalid `fields` argument: `$(kw.second)`.  Must be of the form `(x, y, z)`.")
+            end
+            fields = kw.second
+        else
+            error("$(__source__.file):$(__source__.line): invalid keyword argument for @auto_hash_equals: `$(kw.first)`")
+        end
+    end
+
+    typ = get_struct_decl(__source__::LineNumberNode, typ)
+
+    auto_hash_equals_impl(__source__, typ, fields, cache, hashfn)
+end
+
+function auto_hash_equals_impl(__source__, struct_decl, fields, cache::Bool, hashfn)
+    is_expr(struct_decl, :struct) || error_usage(__source__)
+
+    type_body = struct_decl.args[3].args
+
+    (!cache || !struct_decl.args[1]) ||
+        error("$(__source__.file):$(__source__.line): macro @auto_hash_equals with cached=true should only be applied to a non-mutable struct.")
+
+    (type_name, full_type_name, where_list) = unpack_type_name(__source__, struct_decl.args[2])
+    @assert type_name isa Symbol
+
+    (member_names, member_decls) = get_fields(__source__, struct_decl; prevent_inner_constructors=cache)
+    if isnothing(fields)
+        fields = (member_names...,)
+    else
+        for f in fields
+            f isa Symbol ||
+                error("$(__source__.file):$(__source__.line): invalid field name: `$f`")
+            f in member_names ||
+                error("$(__source__.file):$(__source__.line): field `$f` not found in struct `$type_name`")
+        end
+    end
+
+    base_hash_name = :($Base.hash)
+    if isnothing(hashfn)
+        hashfn = base_hash_name
+    end
+
+    # Add the cache field to the body of the struct
+    if cache
+        push!(type_body, :(_cached_hash::UInt))
+
+        # Add the internal constructor
+        compute_hash = foldl(
+            (r, a) -> :($hashfn($a, $r)),
+            fields;
+            init = :($hashfn($full_type_name)))
+        ctor_body = :(new($(member_names...), $compute_hash))
+        if isnothing(where_list)
+            push!(type_body, :(function $full_type_name($(member_names...))
+                $ctor_body
+            end))
+        else
+            push!(type_body, :(function $full_type_name($(member_names...)) where {$(where_list...)}
+                $ctor_body
+            end))
+        end
+    end
+
+    result = Expr(:block, __source__, esc(:(Base.@__doc__ $struct_decl)), __source__)
+
+    # add function for hash(x, h). hash(x)
+    if cache
+        push!(result.args, esc(:(function $hashfn(x::$type_name, h::UInt)
+                $hashfn(x._cached_hash, h)
+            end)))
+        push!(result.args, esc(:(function $hashfn(x::$type_name)
+                x._cached_hash
+            end)))
```

Comment:
the julia docs indicate that new types should implement the two arg one - is the one arg one needed?

Reply:
It isn't strictly necessary, but becuase it is so much more efficient than the default implementation with one argument for this case, it is worth doing.